Ocamlfilt: Implementation of demangler in pure OCaml

tmcgilchrist opened the pull request:
Adds **ocamlfilt**, a command-line tool that turns mangled OCaml linker symbols back into readable names (equivalent to c++filt or rustfilt). Supporting the flat mangling schemes (both the pre- and post-OCaml-5.3 encodings) and the new structured scheme, with automatic scheme detection. It reads symbols from the command line or stdin.

```shell
$ ocamlfilt _CamlU3FooM3BarF3baz
Foo.Bar.baz
$ nm a.out | ocamlfilt        # or pipe a whole symbol table through it
```

The tool understands the following mangling schemes used in OxCaml/OCaml and detects which one applies per symbol:

- **flat0** — OCaml ≤ 5.2.1: __ separators, $xx hex escapes.
- **flat1** — OCaml ≥ 5.3: the current flat scheme, in both its Linux-like (. separator, \$ xx escapes) and macOS-like (\$ /$$xx/$$$xx) flavours. The two are told apart by a prescan of the symbol, since some byte sequences are ambiguous in isolation.
- **structured** — the new OxCaml structured scheme (_Caml with tagged, length-prefixed path components), which preserves module/class/function nesting, anonymous-function source locations, and inlining markers.

**--format {auto,flat0,flat1,structured}** forces a specific scheme, the default is auto. On macOS the assembler prepends a leading underscore to symbols, and the tool accepts both the bare and underscored forms.
  
Following c++filt, unrecognised input is passed through unchanged and the exit code is always 0.

shym commented:
I’ve rebased and updated this PR. In particular:

- it now matches the current version of #5099, following the changes to the mangling scheme and following as closely as possible the same code structure, with similar function names, to ease the review,
- I simplified the logic by getting rid of `ref`s being passed around and modified in different functions: many functions now return (also) the length of what they parsed,
- demangling failures are now reported, with both a failure message and exit code 1.

I checked that all the symbols in the `ocamlopt` binary from #5099 are demangled without errors and checked on a random sample that the demangling was what I expected. Here is a small sample:

```
_CamlU26Flambda2_types__Type_descrM1TM5DescrF7print_0_25_code
Flambda2_types__Type_descr.T.Descr.print_0

_CamlU5IdentF8fold_allF20partial_fold_data_49_49_code
Ident.fold_all.partial_fold_data_49

_CamlU8TypecoreF8type_letLu23I2e_typecoreml_10135_21Lu22I2e_typecoreml_10251_8F6fn_826_1858_code
Typecore.type_let.fn(typecore.ml:10135:21).fn(typecore.ml:10251:8).fn_826

_CamlU39Flambda2_simplify__Simplify_switch_exprFu18L24_partial_let_42_42_code
Flambda2_simplify__Simplify_switch_expr.partial_let$_42

_CamlU17CamlinternalQuoteIU14Stdlib__FormatF8kfprintfF6fn_260_1170_code
CamlinternalQuote.<inlining>.Stdlib__Format.kfprintf.fn_260

_CamlU25Dwarf_low__Dwarf_operatorM4SizeFu11A3e3e3e__17_64_code
Dwarf_low__Dwarf_operator.Size.>>>_17
```

lthls commented:
Looking at the samples, I'm worried that the last part (the extra number and the "code" string) is dropped.
We usually have distinct symbols for the closure and the code pointer, and to me it looks like you're printing the code pointer as if it were a closure.
Maybe it's not important (the people looking at the demangled names usually don't see closures so there's little confusion), but it feels a bit weird to not be able to distinguish between different specialisations of the same function.

tmcgilchrist committed:
Add an ocamlfilt tool to demangle OCaml symbols

Add a new `ocamlfilt` tool that demangles binary symbols, supporting
the flat mangling schemes (both the pre- and post-OCaml-5.3 encodings)
and the new structured scheme, with automatic scheme detection. It
reads symbols from the command line or stdin.

Co-authored-by: Simon Spies <sspies@janestreet.com>
Co-authored-by: Samuel Hym <samuel@tarides.com>

tmcgilchrist committed:
Add tests documenting suboptimal name mangling

Record the names the structured scheme currently produces for cases it
handles poorly: nested anonymous functions, indistinguishable functor
instances, first-class modules, lazy thunks, and the OOP fragment.
These are a deliberate baseline to improve in follow-on work (e.g.
extending Shape.t and DWARF to cover objects).

tmcgilchrist committed:
Add end-to-end ocamlfilt tests over real compiler output

Compile representative OCaml and OxCaml sources under both the flat and
structured mangling schemes, then demangle every linker symbol with
each ocamlfilt format and diff the resulting table against a committed
reference. The OxCaml source additionally covers -for-pack and
comprehensions. The harness normalises the macOS leading underscore
and pins LC_ALL=C so a single reference works on Linux and macOS.

Co-authored-by: Simon Spies <sspies@janestreet.com>

tmcgilchrist committed:
Add unit tests for ocamlfilt symbol demangling

Focused tests for each scheme (flat0 for <= 5.2, flat1 for >= 5.3, and
structured) plus a rejection test covering malformed and non-OCaml
symbols.

Co-authored-by: Simon Spies <sspies@janestreet.com>

tmcgilchrist committed:
Add cross-module inlining demangling tests

Exercise -O3 cross-module inlining: surviving anonymous closures,
transitive (multi-hop) inlining, functor-defined functions, and
first-class modules. Documents the structured <inlining> marker and
that only the first debuginfo frame is kept (#5099), so intermediate
units drop out of the demangled name.

Co-authored-by: Simon Spies <sspies@janestreet.com>

spiessimon commented on tools/ocamlfilt.ml:
```diff
@@ -0,0 +1,530 @@
+  (** {1 Identifier decoding} *)
+
+  let unbase26 str pos =
+    let rec aux n p =
```
While I think the nesting here is nice for readability, I think there is a risk that code like this allocates one closure per, for example, `pos`. Maybe this should be factored into a module with a signature that clarifies which of these can be used externally.

spiessimon commented on tools/ocamlfilt.ml:
```diff
@@ -0,0 +1,530 @@
+    let esc_pos = ref 0
+    and raw_pos =
+      ref
+        (try String.index sym '_' + 1
```
Can we pull this out of the ref declaration?

spiessimon commented on tools/ocamlfilt.ml:
```diff
@@ -0,0 +1,530 @@
+
+  (* Format anonymous location from filename_line_col to
+     prefix(filename:line:col) *)
+  let format_anonymous_location prefix loc =
```
How about a semi-structured intermediate representation that would separate demangling from pretty-printing? Basically the same as a mangling path in the mangling scheme. (We can probably not recover all the information in there, but I would imagine still a decent chunk (e.g., which kind of path component it is, the location information for an anonymous function, etc). That will simplify changes to the pretty printing when the underlying information doesn't change (and makes those kinds of changes obvious).

tmcgilchrist replied:
I'll write a version that does things that way and see how it looks.

spiessimon commented on tools/ocamlfilt.ml:
```diff
@@ -0,0 +1,530 @@
+  else None
+
+(* Main program *)
+type demangle_format =
```
When would one want to use the demangler with anything other than `auto`? Maybe there should instead be a verbose flag that would say which scheme was used.

tmcgilchrist replied:
There is ambiguity between Flat0 and Flat1 when you consider cross-platform support. *IF* we drop the Flat1 scheme then you could safely default to auto as the prefix will tell which scheme to use.

spiessimon replied:
Oh, that's unfortunate. Is it still ambiguous even if the platform that we are currently demangling for is known?

spiessimon commented on tools/ocamlfilt.ml:
```diff
@@ -0,0 +1,530 @@
+    | _ -> invalid_arg (Printf.sprintf "Cannot decode hexadecimal digit: %c" c)
+end
+
+(* OCaml 5.3+ flat demangling. Trunk emits one of two styles per
```
Could you summarize the differences between Flat0 and Flat1 somewhere? Would it perhaps even be possible to consolidate them into a single demangling pass that locally decides which one is being used?

tmcgilchrist replied:
The key difference is upstream OCaml changed the separator from "__" to "." to fix name clashes with duplicate symbols. This flowed into issues ([#13050](https://github.com/ocaml/ocaml/pull/13050), [#14415](https://github.com/ocaml/ocaml/issues/14415), [#12954](https://github.com/ocaml/ocaml/pull/12954)) with Windows and macOS symbol names needing different separators for their assemblers and debuggers. 

OxCaml hasn't taken that separator change so we could drop `Flat1` entirely if we wanted.

spiessimon reviewed (commented):
I've made a first pass over the actual implementation of the demangling logic.

spiessimon commented:
Based on the tests, I've had Claude assemble some rough edges in the current mangling scheme. I figured it would be useful to surface them here explicitly. We don't have to fix any of them in this PR, and it's not even clear we have to fix all of them.

To structure things a little, I've tried to define goals of a name mangling scheme:

1. **source-identifying**: the path-shaped part of the demangled name (ignoring trailing stamps) determines exactly one source entity;
2. **human-readable**;
3. **path-unambiguous**: the demangled string parses one way — the shape can be read off the syntax;
4. **minimal**;
5. **injective at the symbol level**: no two symbols collide.

Each item below cites a concrete example and flags which goal is hurt. Goal 5 is essential but never violated here, because it is already guaranteed by the stamps appended at the end of each mangled name, so it is listed here only for completeness.

### 1. Redundant location chains in nested anonymous functions

Each enclosing `fun` adds its own `fn(file:line:col)` segment, even though one location uniquely identifies the closure.

- `e2e_suboptimal.reference:6`
  `E2e_suboptimal.nested_lambdas.fn(e2e_suboptimal.ml:33:2).fn(e2e_suboptimal.ml:34:4).fn_N_N_code`
  Source `e2e_suboptimal.ml:32-35`: three nested `fun[@inline never]`.
  The file name is repeated in every link of the chain, and the leaf
  drops its location entirely (asymmetric).
- `e2e_inline.reference:20`
  `E2e_inline.<inlining>.E2e_inline_lib.chain4.fn(e2e_inline_lib.ml:50:2).fn(e2e_inline_lib.ml:51:4).fn_N_N_code`
  The file `e2e_inline_lib.ml` is already implied by the
  `E2e_inline_lib` path component, but it is spelled out again inside
  each `fn(...)`.

Violates **(4)** (minimality). With the `<unit>` prefix in scope, the embedded filename duplicates the same information.

### 2. Functor application sites are erased

The two applications of `Make` in `e2e_suboptimal.ml` produce two distinct compiled copies of `run`, but both mangle to the same name.

- `e2e_suboptimal.reference:21-23` — three rows all decoded to
  `E2e_suboptimal.Make.run_N_N_code`. One is the body of `Make`,
  the others are `Int_inst.run` and `Str_inst.run`. The application
  site `Int_inst` / `Str_inst` is dropped.
- `e2e_ocaml.structured.reference:58-59,60-61` —
  `E2e_ocaml.Make_set.add_N_N_code` and `.mem_N_N_code` each appear
  twice; the functor instance (`Int_set` in source) is invisible.
- `e2e_inline.reference:27`
  `E2e_inline.<inlining>.E2e_inline_lib.Scaler.scale_all_N_N_code`
  comes from `Scaler3 = Scaler(struct ... end)` (source
  `e2e_inline.ml:87-89`); the path says `Scaler`, not `Scaler3`.

Violates **(1)**: the path `Make.run` doesn't pin a single source entity — it can mean the functor body or any application of it.

### 3. First-class modules collapse

Two different `(module struct ... end)` bodies cannot be told apart.

- `e2e_suboptimal.reference:8-9` — both branches of `pick` (source
  `e2e_suboptimal.ml:75-83`, distinct `let v x = x+1` vs `x-1`)
  demangle identically to `E2e_suboptimal.pick.v_N_N_code`.

Violates **(1)**.


### 4. Trailing `_code` and the second stamp are noise in the demangled form

Every code symbol ends with `_<stamp>_<stamp>_code`. The first stamp is genuinely useful — it disambiguates separate copies of the same source function (e.g. specialized versions). The other two pieces don't need to surface in the display:

- `_code` marks that the symbol is the entry code rather than the
  closure data; that distinction matters in the object file but not
  in a demangled path that already shows a function identity.
- The second stamp is there for symbol-level uniqueness (goal
  **(5)**), so it has to stay in the mangled name — but goal (5)
  lives at the symbol level, not in the display, and the demangler
  is free to elide it.

Visible in essentially every line of `e2e_ocaml.structured.reference`, `e2e_oxcaml.plain.reference`, `e2e_inline.reference`, `e2e_suboptimal.reference`.

Violates **(4)**. Both pieces can be dropped from the demangled output without disturbing the underlying symbols.

### 5. Lazy thunks have no path marker

The body of `lazy (...)` is lifted to its own closure, but no `lazy` path item is emitted; it appears as a bare top-level `fn` with no location.

- Source `e2e_suboptimal.ml:92-94` (`deferred`) creates a lazy thunk.
  The corresponding symbol is one of the many indistinguishable
  `E2e_suboptimal.fn_N_N_code` rows
  (`e2e_suboptimal.reference:10-18`).

Violates **(1)**: the path gives no hint that this is a lazy thunk or which one, so it doesn't pin a source entity.

### 6. Class methods all degenerate to anonymous `fn`

- `e2e_suboptimal.reference:10-18` — nine rows of
  `E2e_suboptimal.fn_N_N_code`. Source `e2e_suboptimal.ml:118-135`
  defines methods `sound`, `describe`, `scale`, `rename`, `log`,
  `total`; none of those names survive in the symbol. A method
  `scale` on `dog` is indistinguishable from the top-level `scale`
  (source `e2e_suboptimal.ml:116`) and from any other anonymous
  function in the unit.
- `e2e_suboptimal.reference:27` `E2e_suboptimal.dog.fn_N_N_code` —
  the only method that gets *any* class context, and even then the
  method name is missing.
- `e2e_ocaml.structured.reference:72-73`
  `E2e_ocaml.point2d.fn_N_N_code`,
  `E2e_ocaml.circle.fn_N_N_code` — same pattern: class is named,
  method is not.

Violates **(1)** (the method name is dropped, so the path no longer pins a source entity) and **(2)**.


### 7. The packed unit name contains a redundant copy of the inner unit

Compiling `e2e_oxcaml.ml` with `-for-pack Ox_pack` produces unit symbols whose U-payload is `E2e_oxcamlOx_pack__E2e_oxcaml` (29 chars), i.e. the inner unit name (`E2e_oxcaml`) appears twice with the pack prefix sandwiched between them.

- `e2e_oxcaml.packed.reference:2`
  `E2e_oxcamlOx_pack__E2e_oxcaml.id_succ_N_N_code` (structured).
- `e2e_oxcaml.packed.reference:35`
  `E2e_oxcamlOx_pack.E2e_oxcaml` (flat — at least it splits `__`
  back into `.`).

Violates **(2)** and **(4)**. Also worth noting: the structured scheme leaves the `__` separator inside the U-payload as a literal substring; the flat scheme is the only one that turns it back into `.`.

### 8. Inlined first-class module body reads like a function-local value

When a `pack_op`-style function returns `(module struct let apply x = ...`, the inlined `apply` is attributed under the *function* path, not under any module structure.

- `e2e_inline.reference:22`
  `E2e_inline.<inlining>.E2e_inline_lib.pack_op.apply_N_N_code`.
  Source `e2e_inline_lib.ml:61-64`: `apply` lives inside `(module
  struct ... end)` returned by `pack_op`. The path makes it look
  like `apply` is a direct child of `pack_op`, not of an anonymous
  first-class module.

Violates **(2)**.

### 9. Anonymous-function symbols at unit level are wholly opaque

A bare top-level `fn(...)` without a parent name and without a location collapses many distinct functions into one identical display.

- `e2e_ocaml.structured.reference:12-18` — seven rows of
  `E2e_ocaml.fn_N_N_code`.
- `e2e_oxcaml.plain.reference:13-16` — four rows.
- `e2e_suboptimal.reference:10-18` — nine rows (mix of lazy thunk,
  class methods, and toplevel `fun`s).

Violates **(1)** (one demangled path covers many distinct source closures of unrelated kinds).

### 10. `fn` location appears or vanishes inconsistently

Some closures get a `fn(file:line:col)` segment, some get a stamped `fn` (no location), and they sit next to each other in the same table.

- `e2e_ocaml.structured.reference:21`
  `Alternating.outer_named.middle_named.fn(e2e_ocaml.ml:255:6).inner_named.fn_N_N_code`
  vs `e2e_ocaml.structured.reference:19,20`
  `Alternating.anon_first.named1.fn_N_N_code`,
  `Alternating.outer_named.middle_named.fn_N_N_code` — same source
  unit, same kind of nesting, some lambdas keep their location and
  some do not. The choice is governed by inlining/closure decisions
  rather than the source shape.

Violates **(2)** (the reader cannot predict which closure will be located from the source alone) and **(1)** (the located vs. non-located forms cover different source closures with no documented rule).

### 11. Functor body double-names itself

The body of a functor `F` needs a symbol distinct from anything inside it, so the body has to get *some* name; the current choice is to reuse the functor's own name, which produces `F.F` paths.

- `e2e_ocaml.structured.reference:57`
  `E2e_ocaml.Make_set.Make_set_N_N_code`
- `e2e_ocaml.structured.reference:69`
  `E2e_ocaml.Make_pair.Make_pair_N_N_code`
- `e2e_ocaml.structured.reference:34`
  `E2e_ocaml.Functor_with_lambdas.Functor_with_lambdas_N_N_code`
- `e2e_oxcaml.plain.reference:25,26,20`
  `Bump.Bump`, `Pipeline.Pipeline`, `Make_adder.Make_adder`
- `e2e_suboptimal.reference:24` — `Make.Make`

Violates **(4)** mildly. A dedicated marker would convey the same uniqueness without repeating the functor name.

### 12. The "structured" output is partly flat in practice

Compiling with `-name-mangling-scheme structured` does *not* push every symbol through the structured encoding. Some of the mangled symbols remain flat. A particular source for potential confusion is that entries in the data section are often mangled flat, while the code for the same function is mangled structured.

- `e2e_ocaml.structured.reference:57` (code, structured)
  `_CamlU9E2e_ocamlM8Make_setF10Make_set_N_N_code` ↦
  `E2e_ocaml.Make_set.Make_set_N_N_code`, versus
  `e2e_ocaml.structured.reference:83` (data, flat)
  `camlE2e_ocaml__Make_set_N` ↦ `E2e_ocaml.Make_set_N` — same
  source entity (the `Make_set` functor), two encoders, two
  demangled paths.
- `e2e_ocaml.structured.reference:74-137` — the lower half of the
  table is flat-mangled across the board: operators (`>>=`, `>>|`,
  `.%()`, `.%[]`, `.%{}`, `.%()<-`, `and*`), per-closure `_N`
  stamps, and every functor-module constructor.

Violates **(2)** and **(3)**: a reader has to know in advance which encoder produced any given symbol before they can interpret the path, and the two encoders can map the same source entity to different demangled strings.

spiessimon commented on tools/ocamlfilt.ml:
```diff
@@ -0,0 +1,530 @@
+          (format_anonymous_location "partial" (decode_next ()))
+      | 'I' ->
+        (* Inline_marker *)
+        Buffer.add_string result "<inlining>"
```
I think we should call this `<specialization_of>`, since the actual function has not been inlined here, it has been copied over and potentially specialized.

tmcgilchrist replied:
Can we tell the difference between an inlining and specialization based on the debug_info that we get for a symbol?

spiessimon replied:
The debug info indeed contains the stack traces for inlined functions. 

The symbols that we see here should not actually be functions that have been inlined in the usual sense of inlining, since inlining f into g does not require copying f as a function symbol. Instead, I think these are functions that were copied and then either specialized from their source file or they had to be copied, because they are a closure being defined inside of a function that is being inlined.

tmcgilchrist committed:
Rename <inlining> to <specialization_of>

tmcgilchrist committed:
Extract ref initial_raw_pos

tmcgilchrist committed:
Split demangling into data type and pretty printing

tmcgilchrist committed:
Extract common path_item between mangle/demangle.

spiessimon commented on .github/workflows/build.yml:
```diff
@@ -235,15 +235,15 @@ jobs:
             build_ocamlparam: 'reaper=1,_'
             ocamlparam: 'reaper=1,_'
-            disable_testcases: 'testsuite/tests/tool-ocamlobjinfo/question.ml testsuite/tests/flambda2/examples/array_kind.ml testsuite/tests/flambda2/examples/code_id_join_b.ml testsuite/tests/flambda2/examples/functor_call.ml testsuite/tests/flambda2/examples/invalid.ml testsuite/tests/flambda2/examples/value_slot_c.ml testsuite/tests/flambda2/examples/value_slot_use_b.ml'
+            disable_testcases: 'testsuite/tests/tool-ocamlobjinfo/question.ml testsuite/tests/flambda2/examples/array_kind.ml testsuite/tests/flambda2/examples/code_id_join_b.ml testsuite/tests/flambda2/examples/functor_call.ml testsuite/tests/flambda2/examples/invalid.ml testsuite/tests/flambda2/examples/value_slot_c.ml testsuite/tests/flambda2/examples/value_slot_use_b.ml testsuite/tests/tool-o ...
```
Why do we need to disable these here? With the flag, shouldn't we be able to run these nevertheless?  If there is some instability in the symbols, I suggest documenting in the tests why we have disabled them in case we want to enable them later down the line.

shym replied:
I’ve restricted the set of disabled tests to only the ones that actually fail in 98d0861ecd6661f77c438882e88170bb5a801793 where the commit message states what fails ([logs](https://github.com/shym/oxcaml/actions/runs/26956867381) of the failures, in case you’re interested).
I didn’t understand precisely how you suggest we document the reason for disabling them, as I didn’t find why some other tests chosen at random were disabled (I was wondering if I could find a model on how to do this).

spiessimon replied:
It's fine, we can keep them as is. I was thinking a comment in the .ml file that we disable for why it doesn't work in certain configurations.

spiessimon commented on lambda/debuginfo.ml:
```diff
@@ -46,7 +46,8 @@ module Scoped_location = struct
                assume_zero_alloc: ZA.Assume_info.t;
-               mangling_item: Structured_mangling.path_item option}
+
+      mangling_item: Compilation_unit.t Structured_mangling.path_item option}
```
The formatting here is off, I think.

spiessimon commented on utils/structured_mangling.ml:
```diff
@@ -224,7 +224,204 @@ let mangle_path_item buf path_item =
+  (** Inverse of {!encode}: decode a single length-prefixed identifier at
+      [pos] in [str], returning the decoded string and the number of bytes
+      consumed. *)
+  let decode str pos =
```
Since encoding and decoding can get quite subtle, can we add some tests that they roundtrip? I don't mean end to end demangling tests. Just some tests that use tricky strings (empty, non-ascii characters, etc.) that they encode and then decode again and check that the result is the same as the input.

shym replied:
I’ve changed `Structured_mangling` so that the `decode` and `encode` functions are exported so that they can be tested in `ocamlfilt`. I wondered whether that was too intrusive (I’ve added documentation to `Structured_mangling` to state that they are exported only for testing purposes, but still), so kept it in a separate commit.
Had you another idea on how to test that in a less intrusive way?
I’ve also added a `roundtrip` test. Note that it doesn’t test for the empty string, as the decoder explicitly rejects them. Should we relax the test that ensures the length is positive (I don’t see the use case)?

spiessimon replied:
This is perfectly fine. I think we definitely need to expose these, and that's perfectly fine. I was initially thinking of something like unit tests instead of adding a separate argument to ocamlfilt, but this is also okay.

spiessimon commented on utils/structured_mangling.ml:
```diff
@@ -224,7 +224,204 @@ let mangle_path_item buf path_item =
+        | _ -> first, second, count
+      else find_underscores (i - 1) count first second
+    in
+    let first, second, count = find_underscores (len - 1) 0 (-1) (-1) in
```
I think the loop here is correct, but also very subtle (going from right to left, and turning first into second when we hit `1`). I think we could just use `rindex_from` to save us the trouble of manually implementing this loop.

shym replied:
I was doing a similar edit at that same time :-D

spiessimon commented on utils/structured_mangling.ml:
```diff
@@ -224,7 +224,204 @@ let mangle_path_item buf path_item =
+             if !pos < len then String.sub sym !pos (len - !pos) else ""
+           in
+           Some (List.rev !items, suffix)
+       with Exit | Invalid_argument _ -> None)
```
I think we should avoid the big try catch here. There is probably a more elegant way to exit the loop and deal with the error cases more gracefully.

spiessimon commented on utils/structured_mangling.ml:
```diff
@@ -224,7 +224,204 @@ let mangle_path_item buf path_item =
+       with Exit | Invalid_argument _ -> None)
+end
+
+let mangle_ident (cu : Compilation_unit.t) (path : Compilation_unit.t path) =
```
Shouldn't we put this module at the end of the file?

spiessimon commented on tools/ocamlfilt.ml:
```diff
@@ -0,0 +1,358 @@
+      then Some plen
+      else None
+    in
+    match try_prefix alternate_caml_prefix with
```
Probably the caml prefix is more likely, so it would make sense to swap the two.

spiessimon commented on tools/ocamlfilt.ml:
```diff
@@ -0,0 +1,358 @@
+    | Linux_like
+    | Macosx
+
+  (* A bare ['.'] or ["__"] is a positive Linux-like marker (macOS
```
What is a "positive" marker?

shym replied:
How about “A bare ['.'] or ["__"] can happen only in the Linux-like style…”?

spiessimon replied:
Sounds good!

spiessimon commented on tools/ocamlfilt.ml:
```diff
@@ -0,0 +1,358 @@
+      (* Same rationale as [Flat1.unmangle]: catch the
+         [Invalid_argument] from out-of-bounds index accesses on a
+         truncated symbol, propagate everything else. *)
+      with Invalid_argument _ -> None
```
I think we should avoid the large `Invalid_argument` try-with block and scope it smaller instead.

spiessimon commented on tools/ocamlfilt.ml:
```diff
@@ -0,0 +1,358 @@
+        Some (Bytes.extend result 0 (!j - len) |> Bytes.to_string)
+      (* Out-of-bounds index accesses on a truncated symbol raise
+         [Invalid_argument]; everything else should propagate. *)
+      with Invalid_argument _ -> None
```
I think also here we should avoid the large try-with block.

spiessimon commented on tools/ocamlfilt.ml:
```diff
@@ -0,0 +1,358 @@
+              j := !j + 1;
+              i := !i + 3)
+            else if
+              Char.equal str.[!i] '_' && Char.equal str.[!i + 1] '_'
```
I think there may be a bug here. If a name ends with just an underscore (e.g., `camlFoo_`), this condition will raise an invalid access and, subsequently, the function will return None.

shym replied:
Indeed! I added a bound check before checking the character itself.

spiessimon commented on tools/ocamlfilt.ml:
```diff
@@ -0,0 +1,358 @@
+      let result = Bytes.create len in
+      try
+        while !i < len do
+          if str.[!i] == '_' && str.[!i + 1] == '_'
```
Should probably also be `Char.equal`

spiessimon commented on tools/ocamlfilt.ml:
```diff
@@ -0,0 +1,358 @@
+              j := !j + 1;
+              i := !i + 1)
+        done;
+        Some (Bytes.extend result 0 (!j - len) |> Bytes.to_string)
```
`!j - len` is kind of odd, should we be using `Bytes.sub_string` instead here?

spiessimon reviewed (changes requested):
I took a closer look now, and I think this is largely good at this point. I've mostly found a couple of smaller things. I've intentionally not looked for improvements to the demangling scheme (see the previous comment), since we can deal with those in separate PRs.

shym committed:
Address review

* Simplify `Structured_mangling.parse_location`
* Use consistent exceptions in structured-mangling parsing
* Add CRs
* Reindent `Structured_mangling`
* Rewrite `Structured_mangling.Parsed.parse` in a more functional style
* Swap `mangle_ident` and `Parsed`
* Make `Ocamlfilt.Flat1.detect_style` more readable
* Swap the prefix detection order
* Fix a comment
* Simplify `Flat*.unmangle` functions
* Desugar the Option monad
* Accept a trailing `_` in Flat* schemes
* Catch exceptions from `decode_split_parts`
* Add new structured tests
* Reformulate Flat1.detect_style explanations
* Reformat ocamlfilt
* Move the exception normaliser inside the loop in unhexes
* Add a test for a flat symbol ending on `_`
* Comment in ocamlfilt about the use of `Invalid_argument`
* Fix formatting
* Rename `Parsed` into `Parse`
* Get rid of a useless `option`
* Reformat ocamlfilt

spiessimon commented on utils/structured_mangling.ml:
```diff
@@ -329,30 +329,18 @@ module Parsed = struct
-        | 1 -> find_underscores (i - 1) 2 i first
-        | _ -> first, second, count
-      else find_underscores (i - 1) count first second
+    let ( let* ) = Option.bind in
```
This is better, but let's just use `Option.bind` here instead of the inline operator overwrite.

spiessimon commented on utils/structured_mangling.ml:
```diff
@@ -261,7 +263,8 @@ module Parsed = struct
     in
-    loop pos
+    try loop pos
+    with Invalid_argument _ -> invalid_arg "non-terminated hexadecimal integer"
```
I think we should move this into the loop, where the actual access is.

shym committed:
Add to `ocamlfilt` the encoder and decoder of identifiers

Add access to the identifier encoders and decoders for structured
mangling in `ocamlfilt`
Add a test that some simple symbols round-trip

spiessimon reviewed (approved):
Thanks for cleaning this up. I think we can merge this now.

shym committed:
Handle exceptions as soon as possible in demanglers

Remove exception handlers over large functions in demangler code so that
no unintended exception (beside the exceptions that meant the input
symbol was not well-formed) is caught by mistake.

With the structured mangling scheme, catch exceptions raised by digit
decoders (`unbase26` and `unhexes`) and check length before invoking
`Buffer.add_substring` so that `decode_split_parts` returns `None` on
ill-formed symbols instead of exceptions.

With flat mangling schemes, expand the domain of the demangling
functions so that they fail returning `None` only when the prefix is not
the expected one and never raise exceptions, by letting not-well-formed
input through. Note that the demangler was already demangling
non-well-formed symbols: with `flat0`, `$g` was demangled as `$g`, while
`$f` would fail to demangle. This change makes `$f` demangle as `$f`.

shym committed:
Disable some tests in CI

Disable `ocamlfilt` tests in specific CI configurations in which the set
of generated symbols change so the expected output is off in the tests

shym commented:
I just rebased to resolve the conflict in the CI workflow and added a last commit that removes the uses of `Invalid_argument` in the demanglers to interrupt processing and handle exceptions as soon as possible instead.

- With the structured mangling scheme, catch exceptions raised by digit decoders (`unbase26` and `unhexes`) and check length before invoking `Buffer.add_substring` so that `decode_split_parts` returns `None` on ill-formed symbols instead of exceptions.

- With flat mangling schemes, expand the domain of the demangling functions so that they fail returning `None` only when the prefix is not the expected one and never raise exceptions, by letting not-well-formed input through. So this changes the previous behaviour on some non-well-formed symbols. Note that the demangler was already succeeding on some of non-well-formed symbols: with `flat0`, `$g` was demangled as `$g`, while `$f` would fail to demangle. This change makes `$f` demangle successfully as `$f`. This is close to the intended behaviour of `ocamlfilt` as a whole, where demangling failures are reported by outputting the original input.

shym commented:
About the issues mentioned in a [previous comment](https://github.com/oxcaml/oxcaml/pull/5100#issuecomment-4508166447):

### 1. Redundant location chains in nested anonymous functions

> The file name is repeated in every link of the chain, and the leaf drops its location entirely (asymmetric).

We had decided to drop explicitly the leaf for uniformity with the case where no debuginfo was available. Is it time to revise that decision?

> With the `<unit>` prefix in scope, the embedded filename duplicates the same information.

Do you suggest we always drop the filenames in locations? Possible drawbacks:

- there’s often a little step to go from one to the other, such as `Stdlib__Scanf` to `scanf.ml`
- in case of specialisation, the source isn’t the first module name in the symbol
- I wonder what would happen when there are directives (`# <line> <file>`).

So I suggest we drop the filenames only when they match completely the compilation unit (ie I would keep `scanf.ml` in the example above), I suppose the previous one on the path (in case of specialisation).

### 2. Functor application sites are erased

I’ve spent quite some time digging and experimenting with the suboptimal example to try and find why the information is missing and where we could recover it. My current understanding:
- the copies of `run` are generated somewhere in the simplifier,
- all the debug info attached to those `run`s contain only one item, as seen by the structured name mangling code,
- when I trace the `DA.denv` in `simplify_let0`, I see the metadata associated with `_CamlU3BobM4MakeF5run_1_9_code` change and contain at the end the scope for its instance.
So I suppose the symbol is generated too early, before the simplifier has decided to specialize it, but I don’t know yet how to fix this.

### 4. Trailing `_code` and the second stamp are noise in the demangled form

The second stamp is unique within a given compilation unit? At least, not the first:

```
_CamlU9E2e_ocamlM20Functor_with_lambdasF6nestedLu21J2e_e2e_ocamlml_138_6F8inner_20_86_code
_CamlU9E2e_ocamlM20Functor_with_lambdasF6nestedLu21J2e_e2e_ocamlml_138_6F8inner_20_100_code
```

lthls commented:
> ### 2. Functor application sites are erased
> 
> I’ve spent quite some time digging and experimenting with the suboptimal example to try and find why the information is missing and where we could recover it. My current understanding:
> 
>     * the copies of `run` are generated somewhere in the simplifier,
> 
>     * all the debug info attached to those `run`s contain only one item, as seen by the structured name mangling code,
> 
>     * when I trace the `DA.denv` in `simplify_let0`, I see the metadata associated with `_CamlU3BobM4MakeF5run_1_9_code` change and contain at the end the scope for its instance.
>       So I suppose the symbol is generated too early, before the simplifier has decided to specialize it, but I don’t know yet how to fix this.

Functors are just functions. If you write `let f x = let g y = x + y in g x`, every invocation of `f` creates a new function `g` but you don't expect all of them to have different names.

If you want a mechanism to reflect the current specialisation context when specialising code, that is doable, we'll probably need to meet to sketch out what we want the result to look like.


> ### 4. Trailing `_code` and the second stamp are noise in the demangled form
> 
> The second stamp is unique within a given compilation unit? At least, not the first:
> 
> ```
> _CamlU9E2e_ocamlM20Functor_with_lambdasF6nestedLu21J2e_e2e_ocamlml_138_6F8inner_20_86_code
> _CamlU9E2e_ocamlM20Functor_with_lambdasF6nestedLu21J2e_e2e_ocamlml_138_6F8inner_20_100_code
> ```

The first stamp is unique when entering Flambda 2 (each stamp corresponds to a single function in the input), the second stamp is unique within the current compilation unit. This scheme makes it very easy to spot which input function led to which output functions.

spiessimon commented:
Can we somehow turn this discussion into something more structured (e.g., Github threads)? We should probably also move it over to #6242.

### 1. Redundant location chains in nested anonymous functions

> We had decided to drop explicitly the leaf for uniformity with the case where no debuginfo was available. Is it time to revise that decision?

Thanks for looking into this one. I'm torn. I think it makes sense to preserve the precise location and to preserve the stamps, since both add information (where to find it in the source and which copy it is). Do you have suggestions for how to demangle in this case to add the extra information?

> Do you suggest we always drop the filenames in locations?

I think we should leave those alone for now. That is, I think the characterisation that there is a certain amount of redundancy here is fair (and it's probably not bad that we have documented it here), but I don't see a uniform solution at the moment. My instinct is that it's probably not too distracting. 

### 2. Functor application sites are erased

> I’ve spent quite some time digging and experimenting with the suboptimal example to try and find why the information is missing and where we could recover it.

> If you want a mechanism to reflect the current specialisation context when specialising code, that is doable, we'll probably need to meet to sketch out what we want the result to look like.

Hmm, I think it would be good to preserve this information in some form. At the same time, I wasn't expecting it to be an invasive change. Would it be any easier to capture the module name `Foo` in `module Foo = Make(Bar)` and use that instead?  It's a long shot, but I suspect that would at least cover the most common cases.  

### 4. Trailing _code and the second stamp are noise in the demangled form

> _CamlU9E2e_ocamlM20Functor_with_lambdasF6nestedLu21J2e_e2e_ocamlml_138_6F8inner_20_86_code
> _CamlU9E2e_ocamlM20Functor_with_lambdasF6nestedLu21J2e_e2e_ocamlml_138_6F8inner_20_100_code

I think this is a good reason to keep both stamps. Should we perhaps still drop the _code in the demangler? We should keep it in the mangled name, but I don't see much reason to keep it in the demangled version.

lthls commented:
> Hmm, I think it would be good to preserve this information in some form. At the same time, I wasn't expecting it to be an invasive change. Would it be any easier to capture the module name `Foo` in `module Foo = Make(Bar)` and use that instead? It's a long shot, but I suspect that would at least cover the most common cases.

In Flambda 2, propagating  `Foo` would be strictly harder (the application itself and the name given to its result are at different places). With a small amount of work we could make the bound name available at the application, and then we're back to the same problem (propagating from the application that triggers a specialisation to the specialised functions).
If we want to use names that mean something to the programmer, `Foo` is more likely to retain its original name after compilation to Flambda than `Bar` , so I think this is something we should consider even though it does not make the problem easier to solve.

spiessimon commented on utils/structured_mangling.ml:
```diff
@@ -241,3 +241,194 @@ let mangle_ident (cu : Compilation_unit.t) (path : path) =
+
+module Parse = struct
+  let is_digit = function '0' .. '9' -> true | _ -> false
+  (* CR shym Replace with Char.Ascii.is_digit when the transition to 5.4 is
```
5.4 is merged into main. I think we should be able to do this now.

shym committed:
Merge branch 'main' into ocamlfilt

shym committed:
Use `Char.Ascii` functions for demangling

Drop hand-rolled versions of `is_digit`, `is_hex_digit` and
`hex_digit_to_int` now they are available in `Char.Ascii`

shym commented:
Updated with the current version of `main`, taking advantage of new `Char.Ascii` functions there to avoid duplicates in demangling code.

shym commented:
> Can we somehow turn this discussion into something more structured (e.g., Github threads)?

If there’s any better mechanism, let’s use it indeed! This is clearly poorly suited, I agree.

@lthls: thanks for the explanations. I still wonder about the fact that, when I trace the downwards environments seen by the `let` simplifier, I see many versions of the metadata attached to the `...run_1_9_code` function, and in some of those versions, the symbol we would get would be `...Str_inst.run`. It doesn’t mean it’s easy to get it, but that it’s frustrating not to get it :-)

shym committed:
Merge remote-tracking branch 'ox/main' into ocamlfilt-3

shym committed:
Disable some `ocamlfilt` tests in CI

Disable some CI tests in which the set of symbols changes due to a
different optimisation setting

shym commented:
I updated this PR to merge the changes on the set of CI tests on `main` and, hopefully, disable all the CI tests where the set of symbols changes and so do the test results.

shym commented:
The ocamlformat error on tools/ocamlfilt.ml line 324 on a4465e1 is due to a known [ocamlformat bug](https://github.com/ocaml-ppx/ocamlformat/issues/2771).

spiessimon commented:
I think we should still fix it (e.g., by slightly changing the error message) and the other errors. Could you fix these, @shym?

tmcgilchrist committed:
Run ocamlformat manually
